src/main/kotlin/com/exactpro/th2/codec/ThreadSafeCodec.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2020-2020 Exactpro (Exactpro Systems Limited)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.th2.codec
+
+import com.exactpro.sf.common.messages.structures.IDictionaryStructure
+import com.exactpro.th2.codec.api.IPipelineCodec
+import com.exactpro.th2.codec.api.IPipelineCodecSettings
+import com.exactpro.th2.common.grpc.MessageGroup
+import java.util.concurrent.ConcurrentHashMap
+import javax.annotation.concurrent.ThreadSafe
+
+@ThreadSafe
+class ThreadSafeCodec(
+    private val dictionary: IDictionaryStructure,
+    private val settings: IPipelineCodecSettings?,
+    private val codecFactory: () -> IPipelineCodec
+) : IPipelineCodec {
+    private val instances = ConcurrentHashMap<Thread, IPipelineCodec>()
```

Comment:
Maybe use ThreadLocal instead of ConcurrentHashMap

Reply:
I cannot use `ThreadLocal` because I need to close all instances afterwards